pilopt/src/lib.rs

```diff
@@ -717,3 +747,108 @@ fn remove_duplicate_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
         .collect();
     pil_file.remove_identities(&to_remove);
 }
+
+/// Identifies witness columns that are directly constrained to be equal to other witness columns
+/// through polynomial identities of the form "x = y" and returns a tuple ((name, id), (name, id))
+/// for each pair of identified columns
+fn equal_constrained<T: FieldElement>(
+    expression: &AlgebraicExpression<T>,
+    poly_id_to_array_elem: &BTreeMap<PolyID, (&String, Option<usize>)>,
+) -> Option<((String, PolyID), (String, PolyID))> {
+    match expression {
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
+            left,
+            op: AlgebraicBinaryOperator::Sub,
+            right,
+        }) => match (left.as_ref(), right.as_ref()) {
+            (AlgebraicExpression::Reference(l), AlgebraicExpression::Reference(r)) => {
+                let is_valid = |x: &AlgebraicReference, left: bool| {
+                    x.is_witness()
+                        && !x.next
+                        && if left {
+                            poly_id_to_array_elem.get(&x.poly_id).unwrap().1.is_none()
```

Comment:
So the lhs cannot be an array element? Why? These things should probably be explained in plain language

Reply:
When we find identities like `x = y`, we want to remove one column and replace all its references with the other. However, if we allowed `x[0] = y`, removing x[0] would break the array x integrity since arrays need all their elements to "function properly". 

Added a small comment explaining this situation. Let me know if it's ok.